src/Lykke.AlgoStore.MatchingEngineAdapter.Abstractions/Services/Listening/StreamWrapper.cs

```diff
@@ -0,0 +1,255 @@
+        public IMessageInfo ReadMessage()
+        {
+            CheckDisposed();
+
+            var buffer = new byte[1];
+            _stream.Read(buffer, 0, buffer.Length);
+
+            return ParseMessage(buffer[0], _stream);
+        }
+
+        /// <summary>
+        /// Sends a message
+        /// </summary>
+        /// <typeparam name="T">The type of the message</typeparam>
+        /// <param name="messageId">The ID of the request this message is a reply to</param>
+        /// <param name="messageType">The message type</param>
+        /// <param name="message">The message to send</param>
+        /// <exception cref="ObjectDisposedException">Thrown if the <see cref="StreamWrapper"/> is disposed</exception>
+        public void WriteMessage<T>(uint messageId, byte messageType, T message)
+        {
+            CheckDisposed();
+
+            using (var ms = new MemoryStream())
+            {
+                Serializer.Serialize(ms, message);
+
+                var bytes = ms.ToArray();
+
+                lock (_sync)
+                {
+                    using (var bw = new BinaryWriter(_stream, Encoding.UTF8, true))
+                    {
+                        bw.Write(messageType);
+                        bw.Write(messageId);
+                        bw.Write((ushort)bytes.Length);
+                        bw.Write(bytes);
+                        bw.Flush();
+                    }
+                }
+            }
+        }
+
+        public void MarkAuthenticated()
+        {
+            _isAuthenticated = true;
+        }
+
+        /// <summary>
+        /// Disposes the wrapper and closes the underlying connection
+        /// </summary>
+        public void Dispose()
+        {
+            Dispose(true);
+        }
+
+        protected virtual void Dispose(bool isDisposing)
+        {
+            if (_isDisposed) return;
+
+            if(isDisposing)
+            {
+                _stream.Close();
+                _stream.Dispose();
+            }
+
+            _isDisposed = true;
+        }
+
+        /// <summary>
+        /// Reads a message given a certain message type
+        /// </summary>
+        /// <param name="messageType">The type of the message to read</param>
+        /// <param name="stream">The stream to read the message from</param>
+        /// <returns>A <see cref="MessageInfo"/> containing information about the request and the message</returns>
+        /// <exception cref="InvalidDataException">Thrown when the message type or payload is invalid</exception>
+        private MessageInfo ParseMessage(byte messageType, Stream stream)
+        {
+            if (!_messageTypeMap.ContainsKey(messageType))
+                throw new InvalidDataException($"Message type {messageType} has no handler");
+
+            var result = new MessageInfo(this);
+
+            using (var br = new BinaryReader(stream, Encoding.UTF8, true))
```

Comment:
In case if more then one thread reads from stream from which position begins reading in the line 214?
If BinaryReader resets stream's position to 0, then it corrupts reading. If not, the blocket on the lock thread continues reading from non 0 position.